src/main/java/com/velox/sloan/cmo/workflows/planning/CustomFieldsRetriever.java

```diff
@@ -0,0 +1,464 @@
+
+    @Override
+    public boolean shouldRun() throws RemoteException {
+        return activeTask.getStatus() != activeTask.COMPLETE && activeTask.getTask().getTaskOptions().keySet().contains("DISPLAY CUSTOM FIELDS");
+    }
+
+    @Override
+    public PluginResult run() throws com.velox.api.util.ServerException {
+        try {
+            List<DataRecord> attachedSamples = activeTask.getAttachedDataRecords("Sample", user);
+            List<DataRecord> attachedPlanningProtocols = activeTask.getAttachedDataRecords("PlanningStepProtocol1", user);
+
+            if (!hasValidAttachments(attachedSamples, attachedPlanningProtocols)) {
+                clientCallback.displayError("Task does not have valid attachments or there are no records to display\n" +
+                        "There must be 'Sample' and 'PlanningStepProtocol1' records attached to this task to display custom fields.");
+
+                logError("Task does not have valid attachments or there are no records to display\n" +
+                        "There must be 'Sample' and 'PlanningStepProtocol1' records attached to this task to display custom fields.");
+            }
+            setPlanningStepValues(attachedSamples, attachedPlanningProtocols);
+
+        } catch (Exception e) {
+            clientCallback.displayError(String.format("Error while retrieving custom fields for samples:\n %s", e));
+            logError(String.format("Error while retrieving custom fields for samples:\n%s"), e);
+            return new PluginResult(false);
+        }
+        return new PluginResult(true);
+    }
+
+    /**
+     * The method checks if the active task contains Sample and ProtocolRecord attachments.
+     * @param attachedSamples
+     * @param attachedPlanninProtocols
+     * @return boolean value
+     */
+    private boolean hasValidAttachments(List<DataRecord> attachedSamples, List<DataRecord> attachedPlanninProtocols) {
+        return !(attachedSamples.isEmpty() && attachedPlanninProtocols.isEmpty());
+    }
+
+    /**
+     * This method looks for Sample that has a child record of type IndexBarcode. If not found on the first sample,
+     * the method keep looking upstream the parent hierarchy of the sample until a parent with a child record of IndexBarcode
+     * is found.
+     * @param sample
+     * @return String IndexId value for Sample if found, else returns "" with a warning.
+     * @throws NotFound
+     * @throws RemoteException
+     * @throws IoError
+     */
+    private String getSampleLibraryIndexId(DataRecord sample) throws NotFound, RemoteException, IoError {
+        DataRecord[] indexBarcodeRecord = sample.getChildrenOfType("IndexBarcode", user);
+        String indexId = "";
+        if (indexBarcodeRecord.length > 0) {
+            return indexBarcodeRecord[0].getStringVal("IndexId", user);
+        } else {
+            boolean found = false;
+            DataRecord startingSample = sample;
+            while (!found & startingSample != null) {
```

Comment:
can we use stack, essentially a dfs walking backwards: get rid found , startingSample vairables

Reply:
Created a generic method to get parent samples with required child datatype, and used stacks to track samples